### PR TITLE
Add Foresight Github actions to collect detailed workflow metrics and test results

### PR DESCRIPTION
**Pull request by serkan-ozal**
Hi Otel community,

Thanks for using Foresight to monitor your Github actions. 

Currently OTEL Foresight public view is open to everyone here: [otel.app.runforesight.com](https://otel.app.runforesight.com/)

**Description:**
This PR adds Foresight **workflow-kit** and **test-kit** Github actions into workflows to collect detailed workflow metrics (process traces, CPU/Memory/IO metrics) and test results.

### CODE REVIEW

File: Makefile.Common [hunk 1]
```diff
@@ -19,6 +19,7 @@ LINT=golangci-lint
 IMPI=impi
 # BUILD_TYPE should be one of (dev, release).
 BUILD_TYPE?=release
+RUNNING_ON_GITHUB_ACTION=$(GITHUB_ACTIONS)
```
Comment (jpkrohling) [inline Makefile.Common:22]: There's still something odd here. When running on the main branch, I see the test summary but I don't see it when this PR is checked out:

```
$ make test
go test -race -timeout 300s --tags="" ./...
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter	0.063s
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter/internal/tenant	0.032s
$ gh pr checkout 16427
Switched to branch 'added-foresight'
$ make test
$
```

**Review by jpkrohling — CHANGES_REQUESTED** [review]

File: Makefile.Common (hunk 1 above)
Comment (serkan-ozal) [reply Makefile.Common:22]: Hi @jpkrohling,

There was missing `\` in the `Makefile.common`. Could you please re-pull `added-foresight` branch and re-run?

Comment (jpkrohling) [reply Makefile.Common:22]: I've got the same result. Did you get it working locally? I'm trying it out in the Loki Exporter directory.

Comment (serkan-ozal) [reply Makefile.Common:22]: Ok, sorry for the issue. Let us to reproduce on our local

Comment (serkan-ozal) [reply Makefile.Common:22]: @jpkrohling, we have identified the issue. Working on the fix. FYI

Comment (serkan-ozal) [reply Makefile.Common:22]: @jpkrohling, we have pushed a fix (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16427/commits/b8b62a8743d0268f8a57c998a5154a56151123b7) for silent mode in the Makefile.common. Could you please try again by re-pulling from the branch?

Comment (jpkrohling) [reply Makefile.Common:22]: Looks better, but the output is still different than the one I get from `main`. Looks like it's verbose now?

```
$ make test
if [ "" = "true" ]; then \
	go test -race -v -timeout 300s --tags="" ./... -json > ./foresight-test-report.json; \
else \
	go test -race -v -timeout 300s --tags="" ./...; \
fi
=== RUN   TestLoadConfigNewExporter
=== PAUSE TestLoadConfigNewExporter
=== RUN   TestIsLegacy
=== RUN   TestIsLegacy/not_legacy
=== RUN   TestIsLegacy/format_is_set_to_body
=== RUN   TestIsLegacy/a_label_is_specified
=== RUN   TestIsLegacy/a_tenant_is_specified
=== RUN   TestIsLegacy/a_tenant_ID_is_specified
--- PASS: TestIsLegacy (0.00s)
    --- PASS: TestIsLegacy/not_legacy (0.00s)
    --- PASS: TestIsLegacy/format_is_set_to_body (0.00s)
    --- PASS: TestIsLegacy/a_label_is_specified (0.00s)
    --- PASS: TestIsLegacy/a_tenant_is_specified (0.00s)
    --- PASS: TestIsLegacy/a_tenant_ID_is_specified (0.00s)
=== RUN   TestLoadConfig
=== PAUSE TestLoadConfig
...
...
```

vs.

```
$ make test
go test -race -timeout 300s --tags="" ./...
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter	(cached)
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter/internal/tenant	(cached)
```

Comment (serkan-ozal) [reply Makefile.Common:22]: Ok, because verbose mode is disabled 3 days ago: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16416/files

But we have not rebased from master. Let me rebase from upstream master then, it should be OK.

Comment (serkan-ozal) [reply Makefile.Common:22]: @jpkrohling, Ok, just rebased from upstream master. So verbose mode is disabled for tests. Could you please try again?

Comment (jpkrohling) [reply Makefile.Common:22]: Looks good now!

File: Makefile.Common [hunk 2]
```diff
@@ -54,18 +54,18 @@ common: checklicense impi lint misspell
 
 .PHONY: test
 test:
-	$(GOTEST) $(GOTEST_OPT) ./...
+	$(GOTEST) $(GOTEST_OPT) ./... -json > ./foresight-test-report.json
```
[the remark this answers is not in the document]
Comment (serkan-ozal) [reply Makefile.Common]: Hi @jpkrohling, are you OK to resolve this comment?

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -21,6 +21,11 @@ jobs:
     runs-on: ubuntu-latest
     if: ${{ github.actor != 'dependabot[bot]' }}
     steps:
+      - name: Collect Workflow Telemetry
```
[the remark this answers is not in the document]
Comment (serkan-ozal) [reply .github/workflows/build-and-test.yml:24]: Hi @jpkrohling @mx-psi,

Unfortunately Github actions doesn't support YAML anchors yet. There are lots of request for anchor support from community but it is still not supported yet. Here is its open issue: https://github.com/actions/runner/issues/1182

For some cases, it is suggested to use Github composite actions (https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) to prevent duplication in the YAMLs, but unfortunately it is not useful for our case as there is only single workflow-kit actions added to the jobs to be monitored in detail.

Here is also another discussion in Stack Overflow on sharing steps between jobs.

Long story short, unfortunately, I couldn't find a proper way for our case here. Maybe I am missing something, so do you suggest another solution for this?

[the remark this answers is not in the document]
Comment (mx-psi) [reply .github/workflows/build-and-test.yml:24]: If anchors are not an option it's fine for me to keep it like it is then

[the remark this answers is not in the document]
Comment (serkan-ozal) [reply .github/workflows/build-and-test.yml:24]: Ok, so are you OK to resolve this comment @jpkrohling @mx-psi?
Is there any other thing we need to do to merge this PR?

**Comment by serkan-ozal** [comment]
Hi @jpkrohling, @mx-psi,

Do you know why `changelog` check result fails? https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3541687191/jobs/5946469286
What should we do to get it passed?

File: .github/workflows/build-and-test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpkrohling) [reply .github/workflows/build-and-test.yml:24]: Same to me.

**Review by jpkrohling — APPROVED** [review]
I did not test the `make test` output this time again but verified it earlier. I assume this hasn't changed since I last checked. Therefore, LGTM.
